File: voiceengine/src/main/java/com/backdoor/engine/lang/EnWorker.java
package com.backdoor.engine.lang;

import com.backdoor.engine.misc.Action;
import com.backdoor.engine.misc.Ampm;
import com.backdoor.engine.misc.Long;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class EnWorker extends Worker {

    @Override
    protected String[] getWeekdays() {
        return new String[]{"sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"};
    }

    @Override
    public boolean hasCalendar(String input) {
        return input.matches(".*calendar.*");
    }

    @Override
    public String clearCalendar(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String string = parts[i];
            if (string.matches(".*calendar.*")) {
                parts[i] = "";
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    public List<Integer> getWeekDays(String input) {
        int[] array = {0, 0, 0, 0, 0, 0, 0};
        String[] parts = input.split(WHITESPACES);
        String[] weekDays = getWeekdays();
        for (String part : parts) {
            for (int i = 0; i < weekDays.length; i++) {
                String day = weekDays[i];
                if (part.matches(".*" + day + ".*")) {
                    array[i] = 1;
                    break;
                }
            }
        }
        List<Integer> list = new ArrayList<>();
        for (int anArray : array) list.add(anArray);
        return list;
    }

    @Override
    public String clearWeekDays(String input) {
        String[] parts = input.split(WHITESPACES);
        String[] weekDays = getWeekdays();
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            for (String day : weekDays) {
                if (part.matches(".*" + day + ".*")) {
                    parts[i] = "";
                    break;
                }
            }
        }
        parts = clipStrings(parts).split(WHITESPACES);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i].trim();
            if (!part.matches("on") && !part.matches("at"))
                sb.append(" ").append(part);
        }
        return sb.toString().trim();
    }

    @Override
    public long getDaysRepeat(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            if (hasDays(part)) {
                int integer;
                try {
                    integer = Integer.parseInt(parts[i - 1]);
                } catch (NumberFormatException e) {
                    integer = 1;
                } catch (ArrayIndexOutOfBoundsException e) {
                    integer = 0;
                }
                return integer * DAY;
            }
        }
        return 0;
    }

    @Override
    public String clearDaysRepeat(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            if (hasDays(part)) {
                try {
                    Integer.parseInt(parts[i - 1]);
                    parts[i - 1] = "";
                } catch (NumberFormatException ignored) {
                }
                parts[i] = "";
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    public boolean hasRepeat(String input) {
        return input.matches(".*every.*") || hasEveryDay(input);
    }

    @Override
    public boolean hasEveryDay(String input) {
        return input.matches(".*everyday.*");
    }

    @Override
    public String clearRepeat(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            if (hasRepeat(part)) {
                parts[i] = "";
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    public boolean hasTomorrow(String input) {
        return input.matches(".*tomorrow.*") || input.matches(".*next day.*");
    }

    @Override
    public String clearTomorrow(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            if (part.matches(".*tomorrow.*") || part.matches(".*next day.*")) {
                parts[i] = "";
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    public String getMessage(String input) {
        String[] parts = input.split(WHITESPACES);
        StringBuilder sb = new StringBuilder();
        boolean isStart = false;
        for (String part : parts) {
            if (isStart) sb.append(" ").append(part);
            if (part.matches("text"))
                isStart = true;
        }
        return sb.toString().trim();
    }

    @Override
    public String clearMessage(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            if (part.matches("text")) {
                try {
                    if (parts[i - 1].matches("with")) {
                        parts[i - 1] = "";
                    }
                } catch (IndexOutOfBoundsException ignored) {
                }
                input = input.replace(part, "");
                parts[i] = "";
            }
        }
        return clipStrings(parts);
    }

    @Override
    public Action getMessageType(String input) {
        if (input.matches(".*message.*")) return Action.MESSAGE;
        else if (input.matches(".*letter.*")) return Action.MAIL;
        return null;
    }

    @Override
    public String clearMessageType(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            Action type = getMessageType(part);
            if (type != null) {
                parts[i] = "";
                int nextIndex = i + 1;
                if (nextIndex < parts.length && parts[nextIndex].matches("to")) {
                    parts[nextIndex] = "";
                }
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    public Ampm getAmpm(String input) {
        if (input.matches(".*morning.*")) return Ampm.MORNING;
        else if (input.matches(".*evening.*")) return Ampm.EVENING;
        else if (input.matches(".*noon.*")) return Ampm.NOON;
        else if (input.matches(".*night.*")) return Ampm.NIGHT;
        else if (input.matches(".*a m.*")) return Ampm.MORNING;
        else if (input.matches(".*a.m..*")) return Ampm.MORNING;
        else if (input.matches(".*am.*")) return Ampm.MORNING;
        else if (input.matches(".*p m.*")) return Ampm.EVENING;
        else if (input.matches(".*p.m..*")) return Ampm.EVENING;
        else if (input.matches(".*pm.*")) return Ampm.EVENING;
        return null;
    }

    @Override
    public String clearAmpm(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            if (getAmpm(part) != null) {
                parts[i] = "";
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    protected Date getShortTime(String input) {
        Pattern pattern = Pattern.compile("([01]?[0-9]|2[0-3])( |:)[0-5][0-9]");
        Matcher matcher = pattern.matcher(input);
        if (matcher.find()) {
            String time = matcher.group().trim();
            for (SimpleDateFormat format : getHourFormats()) {
                Date date;
                try {
                    date = format.parse(time);
                    if (date != null) return date;
                } catch (Exception ignored) {
                }
            }
        }
        return null;
    }

    @Override
    public String clearTime(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            if (hasHours(part) != -1) {
                int index = hasHours(part);
                parts[i] = "";
                try {
                    Integer.parseInt(parts[i - index]);
                    parts[i - index] = "";
                } catch (Exception ignored) {
                }
            }
            if (hasMinutes(part) != -1) {
                int index = hasMinutes(part);
                try {
                    Integer.parseInt(parts[i - index]);
                    parts[i - index] = "";
                } catch (Exception ignored) {
                }
                parts[i] = "";
            }
        }
        Pattern pattern = Pattern.compile("([01]?[0-9]|2[0-3])( |:)[0-5][0-9]");
        input = clipStrings(parts);
        Matcher matcher = pattern.matcher(input);
        if (matcher.find()) {
            String time = matcher.group().trim();
            input = input.replace(time, "");
        }
        parts = input.split(WHITESPACES);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i].trim();
            if (!part.matches("at")) sb.append(" ").append(part);
        }
        return sb.toString().trim();
    }

    @Override
    protected int getMonth(String input) {
        int res = -1;
        if (input.contains("january")) res = 0;
        else if (input.contains("february")) res = 1;
        else if (input.contains("march")) res = 2;
        else if (input.contains("april")) res = 3;
        else if (input.contains("may")) res = 4;
        else if (input.contains("june")) res = 5;
        else if (input.contains("july")) res = 6;
        else if (input.contains("august")) res = 7;
        else if (input.contains("september")) res = 8;
        else if (input.contains("october")) res = 9;
        else if (input.contains("november")) res = 10;
        else if (input.contains("december")) res = 11;
        return res;
    }

    @Override
    public boolean hasCall(String input) {
        return input.matches(".*call.*");
    }

    @Override
    public String clearCall(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            if (hasCall(part)) {
                parts[i] = "";
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    public boolean isTimer(String input) {
        input = " " + input + " ";
        return input.matches(".*after.*") || input.matches(".* in .*");
    }

    @Override
    public String cleanTimer(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String string = parts[i];
            if (isTimer(string)) {
                parts[i] = "";
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    public String getDate(String input, Long res) {
        long mills = 0;
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String part = parts[i];
            int month = getMonth(part);
            if (month != -1) {
                int integer;
                try {
                    integer = Integer.parseInt(parts[i + 1]);
                    parts[i + 1] = "";
                } catch (NumberFormatException | IndexOutOfBoundsException e) {
                    integer = 1;
                }
                Calendar calendar = Calendar.getInstance();
                calendar.setTimeInMillis(System.currentTimeMillis());
                calendar.set(Calendar.MONTH, month);
                calendar.set(Calendar.DAY_OF_MONTH, integer);
                mills = calendar.getTimeInMillis();
                parts[i] = "";
                break;
            }
        }
        res.set(mills);
        return clipStrings(parts);
    }

    @Override
    public boolean hasSender(String input) {
        return input.matches(".*send.*");
    }

    @Override
    public String clearSender(String input) {
        String[] parts = input.split(WHITESPACES);
        for (int i = 0; i < parts.length; i++) {
            String string = parts[i];
            if (hasSender(string)) {
                parts[i] = "";
                break;
            }
        }
        return clipStrings(parts);
    }

    @Override
    public boolean hasNote(String input) {
        return input.contains("note");
    }

    @Override
    public String clearNote(String input) {
        input = input.replace("note", "");
        return input.trim();
    }

    @Override
    public boolean hasAction(String input) {
        return input.startsWith("open")
                || input.matches(".*help.*")
                || input.matches(".*adjust.*")
                || input.matches(".*report.*")
                || input.matches(".*change.*");
    }

    @Override
    public Action getAction(String input) {
        if (input.matches(".*help.*")) {
            return Action.HELP;
        } else if (input.matches(".*loudness.*") || input.matches(".*volume.*")) {
            return Action.VOLUME;
        } else if (input.matches(".*settings.*")) {
            return Action.SETTINGS;
        } else if (input.matches(".*report.*")) {
            return Action.REPORT;
        } else {
            return Action.APP;
        }
    }

    @Override
    public boolean hasEvent(String input) {
        return input.startsWith("new") || input.startsWith("add") || input.startsWith("create");
    }

    @Override
    public Action getEvent(String input) {
        if (input.matches(".*birthday.*")) {
            return Action.BIRTHDAY;
        } else if (input.matches(".*reminder.*")) {
            return Action.REMINDER;
        } else return Action.NO_EVENT;
    }

    @Override
    public boolean hasEmptyTrash(String input) {
        return input.matches(".*empty trash.*");
    }

    @Override
    public boolean hasDisableReminders(String input) {
        return input.matches(".*disable reminder.*");
    }

    @Override
    public boolean hasGroup(String input) {
        return input.matches(".*add group.*");
    }

    @Override
    public String clearGroup(String input) {
        StringBuilder sb = new StringBuilder();
        String[] parts = input.split(WHITESPACES);
        boolean st = false;
        for (String s : parts) {
            if (s.matches(".*group.*")) {
                st = true;
                continue;
            }
            if (st) {
                sb.append(s);
                sb.append(" ");
            }
        }
        return sb.toString().trim();
    }

    @Override
    public boolean hasToday(String input) {
        return input.matches(".*today.*");
    }

    @Override
    public boolean hasAfterTomorrow(String input) {
        return input.matches(".*after tomorrow.*");
    }

    @Override
    protected String getAfterTomorrow() {
        return "after tomorrow";
    }

    @Override
    protected int hasHours(String input) {
        if (input.matches(".*hour.*") || input.matches(".*o'clock.*")
                || input.matches(".*am.*") || input.matches(".*pm.*")) return 1;
        return -1;
    }

    @Override
    protected int hasMinutes(String input) {
        if (input.matches(".*minute.*")) return 1;
        return -1;
    }

    @Override
    protected boolean hasSeconds(String input) {
        return input.matches(".*second.*");
    }

    @Override
    protected boolean hasDays(String input) {
        return input.matches(".* day.*");
    }

    @Override
    protected boolean hasWeeks(String input) {
        return input.matches(".*week.*");
    }

    @Override
    protected boolean hasMonth(String input) {
        return input.matches(".*month.*");
    }

    @Override
    public boolean hasAnswer(String input) {
        input = " " + input + " ";
        return input.matches(".* (yes|yeah|no) .*");
    }

    @Override
    public Action getAnswer(String input) {
        if (input.matches(".* ?(yes|yeah) ?.*")) {
            return Action.YES;
        }
        return Action.NO;
    }

    @Override
    protected float findFloat(String input) {
        if (input.matches("half")) {
            return 0.5f;
        } else {
            return -1;
        }
    }

    @Override
    protected String clearFloats(String input) {
        if (input.contains("and a half")) {
            return input.replace("and a half", "");
        }
        if (input.contains("and half")) {
            return input.replace("and half", "");
        }
        if (input.contains(" half an ")) {
            return input.replace("half an", "");
        }
        if (input.contains(" in half ")) {
            return input.replace("in half", "");
        }
        return input;
    }

    @Override
    protected float findNumber(String input) {
        float number = -1;
        if (input.matches("zero") || input.matches("nil")) number = 0;
        if (input.matches("one") || input.matches("first")) number = 1;
        if (input.matches("two") || input.matches("second")) number = 2;
        if (input.matches("three") || input.matches("third")) number = 3;
        if (input.matches("four") || input.matches("fourth")) number = 4;
        if (input.matches("five") || input.matches("fifth")) number = 5;
        if (input.matches("six") || input.matches("sixth")) number = 6;
        if (input.matches("seven") || input.matches("seventh")) number = 7;
        if (input.matches("eight") || input.matches("eighth")) number = 8;
        if (input.matches("nine") || input.matches("ninth")) number = 9;
        if (input.matches("ten") || input.matches("tenth")) number = 10;
        if (input.matches("eleven") || input.matches("eleventh")) number = 11;
        if (input.matches("twelve") || input.matches("twelfth")) number = 12;
        if (input.matches("thirteen") || input.matches("thirteenth")) number = 13;
        if (input.matches("fourteen") || input.matches("fourteenth")) number = 14;
        if (input.matches("fifteen") || input.matches("fifteenth")) number = 15;
        if (input.matches("sixteen") || input.matches("sixteenth")) number = 16;
        if (input.matches("seventeen") || input.matches("seventeenth")) number = 17;
        if (input.matches("eighteen") || input.matches("eighteenth")) number = 18;
        if (input.matches("nineteen") || input.matches("nineteenth")) number = 19;
        if (input.matches("twenty") || input.matches("twentieth")) number = 20;
        if (input.matches("thirty") || input.matches("thirtieth")) number = 30;
        if (input.matches("forty") || input.matches("fortieth")) number = 40;
        if (input.matches("fifty") || input.matches("fiftieth")) number = 50;
        if (input.matches("sixty") || input.matches("sixtieth")) number = 60;
        if (input.matches("seventy") || input.matches("seventieth")) number = 70;
        if (input.matches("eighty") || input.matches("eightieth")) number = 80;
        if (input.matches("ninety") || input.matches("ninetieth")) number = 90;
        return number;
    }

    @Override
    public boolean hasShowAction(String input) {
        return input.matches(".*show.*");
    }

    @Override
    public Action getShowAction(String input) {
        if (input.matches(".*birthdays.*")) {
            return Action.BIRTHDAYS;
        } else if (input.matches(".*active reminders.*")) {
            return Action.ACTIVE_REMINDERS;
        } else if (input.matches(".*reminders.*")) {
            return Action.REMINDERS;
        } else if (input.matches(".*events.*")) {
            return Action.EVENTS;
        } else if (input.matches(".*notes.*")) {
            return Action.NOTES;
        } else if (input.matches(".*groups.*")) {
            return Action.GROUPS;
        } else if (input.matches(".*shopping lists?.*")) {
            return Action.SHOP_LISTS;
        }
        return null;
    }

    @Override
    public boolean hasNextModifier(String input) {
        return input.matches(".*next.*");
    }
}

File: voice-engine-ktx/src/main/java/com/backdoor/engine/lang/EnWorker.kt
package com.backdoor.engine.lang

import com.backdoor.engine.misc.Action
import com.backdoor.engine.misc.Ampm
import com.backdoor.engine.misc.ContactsInterface
import org.threeten.bp.LocalDate
import org.threeten.bp.LocalTime
import org.threeten.bp.ZoneId
import java.util.regex.Pattern

internal class EnWorker(zoneId: ZoneId, contactsInterface: ContactsInterface?) :
  Worker(zoneId, contactsInterface) {

  override val weekdays: List<String> = listOf(
    "sunday",
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday",
    "saturday"
  )

  override fun hasCalendar(input: String): Boolean = input.matches(".*calendar.*")

  override fun clearCalendar(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (s.matches(".*calendar.*")) {
          it[index] = ""
          clearAllBackward(it, index - 1, 3, "and", "add", "to")
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun clearWeekDays(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        for (day in weekdays) {
          if (s.matches(".*$day.*")) {
            it[index] = ""
            clearAllBackward(it, index - 1, 1, "and")
            break
          }
        }
      }
    }.clip()
  }

  override fun getDaysRepeat(input: String): Long =
    input.splitByWhitespaces().firstOrNull { hasDays(it) }?.toRepeat(1) ?: 0

  override fun clearDaysRepeat(input: String): String {
    return input.splitByWhitespaces()
      .toMutableList()
      .also {
        it.forEachIndexed { index, s ->
          if (hasDays(s)) {
            ignoreAny {
              s.toInt()
              it[index - 1] = ""
            }
            it[index] = ""
            return@forEachIndexed
          }
        }
      }.clip()
  }

  override fun hasRepeat(input: String): Boolean =
    input.matches(".*every.*") || hasEveryDay(input)

  override fun hasEveryDay(input: String): Boolean = input.matches(".*every ?day.*")

  override fun clearRepeat(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (hasRepeat(s)) {
          it[index] = ""
          clearAllForward(it, index + 1, 1, "days?")
          return@forEachIndexed
        }
      }
    }.clip().splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (s.matches("repeat")) {
          it[index] = ""
          clearAllBackward(it, index - 1, 1, "and")
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun hasTomorrow(input: String): Boolean =
    input.matches(".*tomorrow.*") || input.matches(".*next day.*")

  override fun clearTomorrow(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (s.matches(".*tomorrow.*") || s.matches(".*next day.*")) {
          it[index] = ""
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun getMessage(input: String): String {
    val sb = StringBuilder()
    var isStart = false
    return input.splitByWhitespaces().forEach {
      if (isStart) sb.append(" ").append(it)
      if (it.matches("text")) isStart = true
    }.let {
      sb.toString().trim()
    }
  }

  override fun clearMessage(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (s.matches("text")) {
          ignoreAny {
            if (it[index - 1].matches("with")) {
              it[index - 1] = ""
            }
          }
          it[index] = ""
        }
      }
    }.clip()
  }

  override fun getMessageType(input: String): Action? = when {
    input.matches(".*message.*") -> Action.MESSAGE
    input.matches(".*letter.*") || input.matches(".*e?( |-)?mail.*") -> Action.MAIL
    else -> null
  }

  override fun clearMessageType(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (getMessageType(s) != null) {
          it[index] = ""
          clearAllForward(it, index + 1, 1, "to")
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun getAmpm(input: String): Ampm? = when {
    " $input ".matches(".* morning .*") -> Ampm.MORNING
    " $input ".matches(".* evening .*") -> Ampm.EVENING
    " $input ".matches(".* noon .*") -> Ampm.NOON
    " $input ".matches(".* night .*") -> Ampm.NIGHT
    " $input ".matches(".* a(.| |)m(.| |) .*") -> Ampm.MORNING
    " $input ".matches(".* p(.| |)?m(.| |) .*") -> Ampm.EVENING
    else -> null
  }

  override fun clearAmpm(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (getAmpm(s) != null) {
          it[index] = ""
          clearAllBackward(it, index - 1, 1, "at")
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun getShortTime(input: String?): LocalTime? {
    return input?.let { s ->
      val matcher = Pattern.compile("([01]?[0-9]|2[0-3])( |:)[0-5][0-9]").matcher(s)
      var localTime: LocalTime? = null
      if (matcher.find()) {
        val time = matcher.group().trim()
        for (format in hourFormats) {
          if (ignoreAny {
              localTime = LocalTime.parse(time, format)
              localTime
            } != null) break
        }
      }
      localTime
    }
  }

  override fun clearTime(input: String?): String {
    val pattern = Pattern.compile("([01]?[0-9]|2[0-3])( |:)[0-5][0-9]")
    return input?.splitByWhitespaces()?.toMutableList()?.also {
      it.forEachIndexed { i, s ->
        if (hasHours(s) != -1) {
          val index = hasHours(s)
          it[i] = ""
          ignoreAny {
            it[i - index].toFloat()
            it[i - index] = ""
          }
        }
        if (hasMinutes(s) != -1) {
          val index = hasMinutes(s)
          ignoreAny {
            it[i - index].toFloat()
            it[i - index] = ""
          }
          it[i] = ""
        }
        ignoreAny {
          val f = s.toFloat()
          if (f >= 0f && f < 24f) {
            it[i] = ""
            clearAllBackward(it, i - 1, 1, "(at|on)")
            if (i < it.size - 1) {
              val minutes = it[i + 1]
              val matcher = pattern.matcher("$s:$minutes")
              if (matcher.find()) {
                it[i + 1] = ""
              }
            }
          }
        }
        ignoreAny {
          val matcher = pattern.matcher(s)
          if (matcher.find()) {
            it[i] = ""
            clearAllBackward(it, i - 1, 1, "(at|on)")
          }
        }
      }
    }?.clip()?.let { s ->
      val matcher = pattern.matcher(s)
      if (matcher.find()) {
        val time = matcher.group().trim()
        s.replace(time, "")
      } else s
    } ?: ""
  }

  override fun getMonth(input: String?): Int = when {
    input == null -> -1
    input.contains("january") -> 1
    input.contains("february") -> 2
    input.contains("march") -> 3
    input.contains("april") -> 4
    input.contains("may") -> 5
    input.contains("june") -> 6
    input.contains("july") -> 7
    input.contains("august") -> 8
    input.contains("september") -> 9
    input.contains("october") -> 10
    input.contains("november") -> 11
    input.contains("december") -> 12
    else -> -1
  }

  override fun hasCall(input: String): Boolean = input.matches(".*call.*")

  override fun clearCall(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (hasCall(s)) {
          it[index] = ""
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun hasTimer(input: String): Boolean = input.let { " $it " }.let {
    (it.matches(".*after.*") || it.matches(".* in .*")) && getMonth(it) == -1
  }

  override fun cleanTimer(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (hasTimer(s)) {
          it[index] = ""
          return@forEachIndexed
        }
      }
    }.clip().trim()
  }

  override fun getDateAndClear(input: String, result: (LocalDate?) -> Unit): String {
    var localDate: LocalDate? = null
    return input.splitByWhitespaces().toMutableList().also { list ->
      list.forEachIndexed { index, s ->
        val month = getMonth(s)
        if (month != -1) {
          var dayOfMonth = ignoreAny({
            list[index + 1].toFloat().also { list[index + 1] = "" }
          }) { -1f }
          if (dayOfMonth == -1f) {
            dayOfMonth = ignoreAny({
              list[index - 1].toFloat().also { list[index - 1] = "" }
            }) { 1f }
          }

          var parsedDate = LocalDate.now(zoneId)
            .withDayOfMonth(dayOfMonth.toInt())
            .withMonth(month)

          if (parsedDate.isBefore(LocalDate.now(zoneId))) {
            parsedDate = parsedDate.plusYears(1)
          }

          localDate = parsedDate
          list[index] = ""
          clearAllBackward(list, index - 1, 1, "(on|at)")
          return@forEachIndexed
        }
      }
    }.clip().also {
      result(localDate)
    }
  }

  override fun hasSender(input: String): Boolean = input.matches(".*send.*")

  override fun clearSender(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (hasSender(s)) {
          it[index] = ""
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun hasNote(input: String): Boolean = input.contains("note")

  override fun clearNote(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (s.matches(".*note.*")) {
          it[index] = ""
          clearAllBackward(it, index - 1, 1, "create", "add")
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun hasAction(input: String): Boolean {
    return (input.startsWith("open")
      || input.matches(".*help.*")
      || input.matches(".*adjust.*")
      || input.matches(".*report.*")
      || input.matches(".*change.*"))
  }

  override fun getAction(input: String): Action = when {
    input.matches(".*help.*") -> Action.HELP
    input.matches(".*loudness.*") || input.matches(".*volume.*") -> Action.VOLUME
    input.matches(".*settings.*") -> Action.SETTINGS
    input.matches(".*report.*") -> Action.REPORT
    else -> Action.APP
  }

  override fun hasEvent(input: String): Boolean = input.startsWith("new") ||
    input.startsWith("add") || input.startsWith("create")

  override fun getEvent(input: String): Action = when {
    input.matches(".*birthday.*") -> Action.BIRTHDAY
    input.matches(".*reminder.*") -> Action.REMINDER
    else -> Action.NO_EVENT
  }

  override fun hasEmptyTrash(input: String): Boolean =
    input.matches(".*(empty|clear) (trash|completed reminders?).*")

  override fun hasDisableReminders(input: String): Boolean =
    input.matches(".*disable (all )?reminders?.*")

  override fun hasGroup(input: String): Boolean = input.matches(".*(add|create|new) group.*")

  override fun clearGroup(input: String): String {
    return input.splitByWhitespaces().toMutableList().also {
      it.forEachIndexed { index, s ->
        if (s.matches(".*group.*")) {
          it[index] = ""
          clearAllBackward(it, index - 1, 2, "an?", "(add|create|new)")
          return@forEachIndexed
        }
      }
    }.clip()
  }

  override fun hasToday(input: String): Boolean = input.matches(".*today.*")

  override fun hasAfterTomorrow(input: String): Boolean = input.matches(".*after tomorrow.*")

  override val afterTomorrow: String = "after tomorrow"

  override fun hasHours(input: String?): Int = when {
    input == null -> -1
    input.matches(".*hour.*") || input.matches(".*o'clock.*") || getAmpm(input) != null -> 1
    else -> -1
  }

  override fun hasMinutes(input: String?): Int = when {
    input.matchesOrFalse(".*minute.*") -> 1
    else -> -1
  }

  override fun hasSeconds(input: String?): Boolean = input.matchesOrFalse(".*seconds?.*")

  override fun hasDays(input: String?): Boolean = input.matchesOrFalse(".* days?.*") &&
    !input.matchesOrFalse(".*birthdays?.*")

  override fun hasWeeks(input: String?): Boolean = input.matchesOrFalse(".*weeks?.*")

  override fun hasMonth(input: String?): Boolean = input.matchesOrFalse(".*months?.*")

  override fun hasAnswer(input: String): Boolean =
    input.let { " $it " }.matches(".* (yes|yeah|no) .*")

  override fun getAnswer(input: String): Action = when {
    input.matches(".* ?(yes|yeah) ?.*") -> Action.YES
    else -> Action.NO
  }

  override fun findFloat(input: String?): Float = input?.takeIf { it.matches("half") }
    ?.let { 0.5f } ?: -1f

  override fun clearFloats(input: String?): String? = when {
    input == null -> null
    input.contains("and a half") -> input.replace("and a half", "")
    input.contains("and half") -> input.replace("and half", "")
    input.contains(" half an ") -> input.replace("half an", "")
    input.contains(" in half ") -> input.replace("in half", "")
    else -> input
  }

  override fun findNumber(input: String?): Float = when {
    input == null -> -1f
    input.matches("zero") || input.matches("nil") -> 0f
    input.matches("one") || input.matches("first") -> 1f
    input.matches("two") || input.matches("second") -> 2f
    input.matches("three") || input.matches("third") -> 3f
    input.matches("four") || input.matches("fourth") -> 4f
    input.matches("five") || input.matches("fifth") -> 5f
    input.matches("six") || input.matches("sixth") -> 6f
    input.matches("seven") || input.matches("seventh") -> 7f
    input.matches("eight") || input.matches("eighth") -> 8f
    input.matches("nine") || input.matches("ninth") -> 9f
    input.matches("ten") || input.matches("tenth") -> 10f
    input.matches("eleven") || input.matches("eleventh") -> 11f
    input.matches("twelve") || input.matches("twelfth") -> 12f
    input.matches("thirteen") || input.matches("thirteenth") -> 13f
    input.matches("fourteen") || input.matches("fourteenth") -> 14f
    input.matches("fifteen") || input.matches("fifteenth") -> 15f
    input.matches("sixteen") || input.matches("sixteenth") -> 16f
    input.matches("seventeen") || input.matches("seventeenth") -> 17f
    input.matches("eighteens?") || input.matches("eighteenth") -> 18f
    input.matches("nineteen") || input.matches("nineteenth") -> 19f
    input.matches("twenty") || input.matches("twentieth") -> 20f
    input.matches("thirty") || input.matches("thirtieth") -> 30f
    input.matches("forty") || input.matches("fortieth") -> 40f
    input.matches("fifty") || input.matches("fiftieth") -> 50f
    input.matches("sixty") || input.matches("sixtieth") -> 60f
    input.matches("seventy") || input.matches("seventieth") -> 70f
    input.matches("eighty") || input.matches("eightieth") -> 80f
    input.matches("ninety") || input.matches("ninetieth") -> 90f
    else -> -1f
  }

  override fun hasShowAction(input: String): Boolean = input.matches(".*show.*")

  override fun getShowAction(input: String): Action? = when {
    input.matches(".*birthdays.*") -> Action.BIRTHDAYS
    input.matches(".*active reminders.*") -> Action.ACTIVE_REMINDERS
    input.matches(".*reminders.*") -> Action.REMINDERS
    input.matches(".*events.*") -> Action.EVENTS
    input.matches(".*notes.*") -> Action.NOTES
    input.matches(".*groups.*") -> Action.GROUPS
    input.matches(".*shopping lists?.*") -> Action.SHOP_LISTS
    else -> null
  }

  override fun hasNextModifier(input: String): Boolean = input.matches(".*next.*")

  override fun hasConnectSpecialWord(input: String): Boolean {
    return input.matches("(at|on)")
  }
}
